### ISSUE TITLE
test: Write assumeutxo tests

### ISSUE THREAD

**Issue by maflcko** [bitcoin/bitcoin#28648]
### Motivation

Tests are useful to catch bugs, so they should be written.

### Possible solution

Write tests, similar to https://github.com/bitcoin/bitcoin/pull/28647

Refer to the `TODO` comments in `./test/functional/feature_assumeutxo.py`, or to the uncovered lines in the AssumeUtxo code via a coverage report. For example, https://maflcko.github.io/b-c-cov/total.coverage/src/validation.cpp.gcov.html

### Useful Skills

* Compiling Bitcoin Core from source
* Running the C++ unit tests and the Python functional tests
* Basic AssumeUtxo understanding
* Bitcoin Core serialization and test framework understanding
* Python serialization understanding

### Guidance for new contributors

Want to work on this issue?

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

**Comment by ryanofsky** [comment]
Original review comment might provide more explanation (https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1331521797):

> Test seems to have pretty good coverage for the happy case, but it would be good to have coverage for pathological cases to make sure checks in the code written to handle them work correctly.
> 
> It should be possible to write a test using `itertools.product` or something similar to test different cases and starting states.
> 
> Interesting test cases could be loading an assumeutxo snapshot file with:
> 
> 1. An invalid hash
> 2. Valid hash but invalid snapshot file (bad coin height or truncated file or bad other serialization)
> 3. Valid snapshot file, but referencing an unknown block
> 4. Valid snapshot file, but referencing a snapshot block that turns out to be invalid, or has an invalid parent
> 5. Valid snapshot file and snapshot block, but the block is not on the most-work chain
> 
> Interesting starting states could be loading a snapshot when the current chain tip is:
> 
> 1. An ancestor of snapshot block
> 2. Not an ancestor of the snapshot block but has less work
> 3. The snapshot block
> 4. A descendant of the snapshot block
> 5. Not an ancestor or a descendant of the snapshot block and has more work
> 
> Most combinations of test cases / starting states above should be possible. I don't think a test like this needs to be part of this PR, but would be nice as a followup. It might make sense to include a consolidated TODO comment in this PR documenting ways the test could be improved (also mentioning other test ideas [#27596 (comment)](https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1325243993) [#27596 (comment)](https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1325246788))

**Comment by BrandonOdiwuor** [comment]
@maflcko would like to work on this

**Comment by kashifs** [comment]
I will help review this PR

**Comment by jrakibi** [comment]
Is anyone working on this? I'd be happy to take a look

**Comment by BrandonOdiwuor** [comment]
@jrakibi I'm currently working on https://github.com/bitcoin-core/gui/issues/769, you can work on it

**Comment by Sjors** [comment]
I created a wallet test file as part of #28616. Feel free to lift that commit out and make a fresh pull request. Right now it doesn't actually test the functionality that's introduced in 28616, though I might expand it.

**Comment by mohamedawnallah** [comment]
Hey, @jrakibi Are you still working on this issue? If not I’d like to pick it up. Thanks

**Comment by nafisalawalidris** [comment]
Hello @maflcko, I'm interested in working on this issue and adding tests for the AssumeUtxo feature. I will follow the guidance provided in the issue description and the template for writing tests. I will keep you posted on my progress and submit a pull request once the tests are ready.

**Comment by jrakibi** [comment]
> Hey, @maflcko @jrakibi Is there anyone working on this issue? If not I’d like to pick it up. Thanks

Hey @mohamedawnallah , I am currently preoccupied with other assignments, feel free to take it over

**Comment by hernanmarino** [comment]
> There'd be [#27596 (comment)](https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1344713537) if someone wants to tackle it.

Done, reviews welcome https://github.com/bitcoin/bitcoin/pull/29394

**Comment by fanquake** [comment]
There are more PRs open. i.e #29428.

**Comment by fjahr** [comment]
I think this may be ready for closing since all remaining TODOs have been addressed IMO. There are three remaining PRs that still need to get merged but I think we can expect that nobody will start a new PR based on the TODOs comment anymore.

For reference, the remaining PRs are #30403, #30320, #29996. See also https://github.com/bitcoin/bitcoin/pull/30403#issuecomment-2213758446

**Comment by fjahr** [comment]
Or rather, the problem is that the issue here is not specific and so none of the issues can close this issue because it's unclear which will be the last one to get merged. I will recommend in the PRs that they mention the issue here as related to so it's at least clear what is being worked on where.

**Comment by Edu92337** [comment]
Hi! I've analyzed the current test coverage and found a gap in RPC error testing.

Current coverage:
- loadtxoutset: ~12 error scenarios covered
- dumptxoutset: only success cases tested

Proposed additions:
- dumptxoutset error cases: ~6 scenarios (file exists, invalid params, pruned mode, etc.)
- loadtxoutset additional errors: ~4 scenarios (empty file, permissions, etc.) 

Would adding comprehensive RPC error testing be helpful for this issue?
I can create a PR with these tests.

**Comment by fjahr** [comment]
> Hi! I've analyzed the current test coverage and found a gap in RPC error testing.
> 
> Current coverage:
> 
> * loadtxoutset: ~12 error scenarios covered
> * dumptxoutset: only success cases tested
> 
> Proposed additions:
> 
> * dumptxoutset error cases: ~6 scenarios (file exists, invalid params, pruned mode, etc.)
> * loadtxoutset additional errors: ~4 scenarios (empty file, permissions, etc.)
> 
> Would adding comprehensive RPC error testing be helpful for this issue? I can create a PR with these tests.

These coverage numbers are incorrect. For example there are several error cases tested for `dumptxoutset` in `rpc_dumptxoutset.py`.

**Comment by fjahr** [comment]
> I think this may be ready for closing since all remaining TODOs have been addressed IMO. There are three remaining PRs that still need to get merged but I think we can expect that nobody will start a new PR based on the TODOs comment anymore.
> 
> For reference, the remaining PRs are [#30403](https://github.com/bitcoin/bitcoin/pull/30403), [#30320](https://github.com/bitcoin/bitcoin/pull/30320), [#29996](https://github.com/bitcoin/bitcoin/pull/29996). See also [#30403 (comment)](https://github.com/bitcoin/bitcoin/pull/30403#issuecomment-2213758446)

> Or rather, the problem is that the issue here is not specific and so none of the issues can close this issue because it's unclear which will be the last one to get merged. I will recommend in the PRs that they mention the issue here as related to so it's at least clear what is being worked on where.

I still think this issue should either be closed as completed or reworked to be much more specific. But probably the best path forward is to just close this and open issues for the specific cases that are still missing, if there are any.

**Comment by yashbhutwala** [comment]
> I still think this issue should either be closed as completed or reworked to be much more specific. But probably the best path forward is to just close this and open issues for the specific cases that are still missing, if there are any.

I agree this issue should be closed. After reviewing the full history:

**All items from ryanofsky's original test matrix have been addressed:**

Test cases (loading snapshot with):
- ✅ Invalid hash - covered in `test_invalid_snapshot_scenarios`
- ✅ Invalid snapshot file - covered in `test_invalid_snapshot_scenarios`
- ✅ Unknown block reference - covered in `test_invalid_snapshot_scenarios`
- ✅ Invalid/invalid parent block - covered in `test_snapshot_block_invalidated`
- ✅ Block not on most-work chain - covered via #30320

Starting states:
- ✅ Ancestor of snapshot - normal case throughout tests
- ✅ Not ancestor, less work - covered via #29996
- ✅ Tip is snapshot block - covered via #30403
- ✅ Descendant of snapshot - implicitly covered (fails with "Work does not exceed")
- ✅ Not ancestor/descendant, more work - covered via #30320

**The three remaining PRs mentioned have all been merged:** #30403, #30320, #29996

**Remaining minor item:** I've opened #34286 to address the last TODO in `wallet_assumeutxo.py` (line 242), which adds test assertions verifying node state after restart during background sync. This documents an important edge case: wallets cannot be loaded after a node restart during assumeutxo background sync because blocks before the snapshot are unavailable for rescanning.

Given the original scope has been fully addressed, I think closing this issue makes sense. Any future test coverage improvements can be tracked in more specific issues.

**Comment by pratikshmalik05** [comment]
can i try to do or pr is already done ?
